### PR TITLE
Declarative marking of RTypedData struct references

### PR DESCRIPTION
**Pull request by eightbitraptor**
This PR is a work in progress.

I've added a flag `RUBY_TYPED_DECL_MARKING` that can be used inside the `rb_data_type_t` flags section to determine whether a T_DATA object can be declaratively marked or not.

If this flag is set then we can use the `data` struct member as a pointer to an array of offsets that we can use in conjunction with the `data` struct to find fields that require marking.

I've modified `gc_mark_children` to use this offset array if present to test the system.

### CODE REVIEW

File: dir.c [hunk 1]
```diff
@@ -487,10 +487,14 @@ dir_memsize(const void *ptr)
     return sizeof(struct dir_data);
 }
 
+#define END_REFS UINTPTR_MAX
+
+static VALUE references[] = {offsetof(struct dir_data, path), END_REFS};
```
Comment (chrisseaton) [inline dir.c:492]: This should be `offset_t references[]` shouldn't it?

And, after it works, we should then transition it to be a bitmap.

**Comment by chrisseaton** [comment]
This is the general idea as I understood it, yes.

File: dir.c (hunk 1 above)
Comment (eightbitraptor) [reply dir.c:492]: I think this should probably be `size_t references[]` (`off_t` is part of POSIX, and the docs tell me it's normally used for file sizes).

I assume I didn't get warned about this is that both VALUE and `size_t` are the same size integer types...

Comment (chrisseaton) [reply dir.c:492]: Looks like you're right they should be `size_t`.

**Comment by eightbitraptor** [comment]
I'm going to investigate and try and benchmark this approach on top of Ruby master, so I'm closing this PR for now to avoid having a convoluted history on this fork.

**Pull request closed without merging**